tests/sentry/utils/test_data_scrubber.py

```diff
@@ -411,3 +412,52 @@ def test_breadcrumb_message(self):
         proc.apply(data)
 
         assert data["breadcrumbs"]["values"][0]["message"] == FILTER_MASK
+
+    def test_can_whitelist_breadcrumb_field(self):
+        data = {
+            "breadcrumbs": {"values": [{"message": "I'm full of secrets. Shhhhh, don't tell!"}]}
+        }
+        data2 = deepcopy(data)
+
+        # without whitelisting "message", this gets scrubbed, because of the
+        # presence of the word "secret"
+        proc = SensitiveDataFilter()
+        proc.apply(data)
+        assert data["breadcrumbs"]["values"][0]["message"] == FILTER_MASK
+
+        # if we whitelist "message," though, the data comes through
+        proc2 = SensitiveDataFilter(exclude_fields=["message"])
+        proc2.apply(data2)
+        assert (
+            data2["breadcrumbs"]["values"][0]["message"]
+            == "I'm full of secrets. Shhhhh, don't tell!"
+        )
+
+    def test_can_whitelist_http_field(self):
+        data = {
+            "request": {
+                # Note: intentionally broken query string to get the filter to act
+                # on the entire value (rather than treat it as an actual query string).
+                # This is in order to show a top-level key getting scrubbed without the
+                # whitelist. (Currently, every value which can get scrubbed in the
+                # request is some sort of nested data structure, and you can't blanket-whitelist
+                # nested data (in case it includes blacklist data), but for future-proofing,
```

Comment:
Why can't you put nested structures in the exclude list? I don't see anything that would prevent that usage.

Reply:
Yes, you're right, you _can_ put them there (in the whitelist), but it won't actually do anything/ have any effect on what does or doesn't get scrubbed.

For example, say your event data contains the following:

```
"my_info": {
    "stuff": "things",
    "dogs": "yes",
    "cats": "maybe",
    "secret_stuff": "I should get filtered"
},
"other_info": "my password is 'd0gsAr3Gr3at!'"
```

Even if you whitelist `my_info`, you'll still end up with

```
"my_info": {
    "stuff": "things",
    "dogs": "yes",
    "cats": "maybe",
    "secret_stuff": "[Filtered]"
},
"other_info": "[Filtered]"
```

because we scrub depth-first, so by the time we get to evaluating the key-value pair `("my_info", <value>)`, that value has already been scrubbed, and `"secret_stuff"`'s value has already been redacted. In other words, whitelisting `my_info` is a noop.

On the other hand, if you whitelist `other_info`, you'll get this:

```
"my_info": {
    "stuff": "things",
    "dogs": "yes",
    "cats": "maybe",
    "secret_stuff": "[Filtered]"
},
"other_info": "my password is 'd0gsAr3Gr3at!'"
```

because `"other_info"` is the key for a leaf, which `"my_info"` is not.

The point of the test is to show that top-level keys can be whitelisted (by showing what happens when I did and didn't include the keyname in `exclude_fields`), so I needed a top-level key whose value was a leaf. And because currently every top level key in `request` [that we consider for scrubbing](https://github.com/getsentry/sentry/blob/9b1393d4a3aa5f2a4012fff6fc8546baf73a6f46/src/sentry/utils/data_scrubber.py#L146) corresponds [to a dictionary](https://github.com/getsentry/sentry/blob/9b1393d4a3aa5f2a4012fff6fc8546baf73a6f46/src/sentry/interfaces/http.py#L92-L104) (or something which gets parsed as a dictionary), I had to fudge things a little.

Does that make sense?

Now, given that whitelisting is a moot point for all of the current top-level keys, why have the test at all? Just in case we ever add a key which _does_ correspond directly to a literal value, its nice to know that whitelisting would work. A touch overkill-y? Probably. More confusing than helpful? I'd hoped not, especially with the comment, but if you think it does more harm than good, I'm open to either revising the comment or taking the test out entirely.

Lemme know what you think.

Reply:
It makes sense now. I was unsure of why you couldn't exclude a top level key. But the depth first traversal explains it :+1: